### PR TITLE
Crawler and Analyzer APIs

### PR DESCRIPTION
**Pull request by sundas-arooj**
### Task:
**The crawler part:**

- Select one social network of your choice (Instagram, Twitter, etc.)
- The crawler should return up to 5 images found searched by keyword

**ML part:**

- Choose a very simple ML model of your choice to classify images (also, your choice)
- Accuracy doesn't matter much for this assignment

**The API would have two endpoints:**

- Search via crawler (described below) by keyword on a social network and return ML analysis

    - The flow would be: API request with keyword >> crawler searches and downloads up to 5 images on a social network >> ML model performs analyses over the images >> API responds with results in JSON the request

- Uploading images directly and getting ML analysis for that image

   - The flow would be: API request with image file >> ML model perform analysis over the image >> API responds with results in JSON the request

**Endpoints:**

> GET /search-via-crawler?keyword=

> POST /analyze
> body: file

### CODE REVIEW

File: app.py [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+from config import DEBUG, HOST, PORT
+from api.routes.base import BaseRoutes
+
+app = Flask(__name__)
+
+# Instantiate BaseRoutes class
+routes = BaseRoutes(app)
+
+# Register routes dynamically
+routes.register_routes()
```
Comment (sundas-arooj) [inline app.py:11]: Initially `BaseRoutes` is initialize here and register all the routes by calling this method

File: api/routes/base.py [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+from api.controllers.base import BaseController
+
+ROUTES = [
+    {
+        'endpoint': '/', 
+        'handler': 'get', 
+        'methods': ['GET'],
+        'controller': BaseController
+    }
+]
```
Comment (sundas-arooj) [inline api/routes/base.py:12]: This is the default endpoint which tells that server is running. Each file under routes defined the routes and has its associated Controllers defined

File: api/routes/base.py [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+        self.excluded_files = ['__init__.py']
+
+    def register_routes(self):
+        routes_dir = os.path.join(os.path.dirname(__file__), '')
+        for filename in os.listdir(routes_dir):
+            if filename.endswith('.py') and filename not in self.excluded_files:
+                module_name = f'api.routes.{filename[:-3]}'
+                routes = __import__(module_name, fromlist=['ROUTES']).ROUTES
+                for route in routes:
+                    self.app.add_url_rule(route['endpoint'], view_func=getattr(route['controller'](), route['handler']), methods=route.get('methods', ['GET']))
```
Comment (sundas-arooj) [inline api/routes/base.py:26]: This method will register all the routes we have in the application

File: api/routes/search_and_analyze.py
Comment (sundas-arooj) [inline api/routes/search_and_analyze.py:1]: This file defines the search and analysis endpoint and import its respective controller

File: api/routes/analyze.py
Comment (sundas-arooj) [inline api/routes/analyze.py:1]: This file defines the analyze endpoint and import its respective controller

File: api/controllers/base.py [hunk 4]
```diff
@@ -0,0 +1,11 @@
+from flask import jsonify
+
+class BaseController:
+    def get(self):
```
Comment (sundas-arooj) [inline api/controllers/base.py:4]: This is the base controller which has base endpoint

File: api/controllers/search_and_analyze.py [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+from config import ASSETS_URL
+
+class SearchAndAnalyzeController(BaseController):
+    def __init__(self):
+        super().__init__()
+        self.crawler = Crawler()
+        self.classifier = Classifier()
+        self.common = Common()
+
+    def search_via_crawler(self):
```
Comment (sundas-arooj) [inline api/controllers/search_and_analyze.py:16]: This function will be called on calling the endpoint `GET /search-via-crawler`

File: api/controllers/search_and_analyze.py [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+        keyword = request.args.get('keyword')
+        if not keyword:
+            return self.error_response('Keyword parameter is required.', 400)
+
+        # Get images from Crawler
+        images = self.crawler.search_images(keyword)
```
Comment (sundas-arooj) [inline api/controllers/search_and_analyze.py:22]: It searches for the image on social media by calling the crawler method

File: api/controllers/search_and_analyze.py [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+        if not images:
+            return self.error_response('No images found for the keyword.', 404)
+
+        # Perform ML analysis on images
+        analysis_results = []
+        for i, image in enumerate(images):
+            current_time = datetime.now().strftime('%Y-%m-%d_%H-%M-%S')
+            file_path = f'{ASSETS_URL}/{keyword}_{i+1}_{current_time}.jpg'
+
+            self.common.download_image(image, file_path)
```
Comment (sundas-arooj) [inline api/controllers/search_and_analyze.py:32]: Here we download the images we get from crawler

File: api/controllers/search_and_analyze.py [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+
+            analysis = self.classifier.analyze_image(file_path)
```
Comment (sundas-arooj) [inline api/controllers/search_and_analyze.py:34]: We perform image classification on those images

File: api/services/crawler.py [hunk 9]
```diff
@@ -0,0 +1,28 @@
+from bs4 import BeautifulSoup
+import selenium.webdriver as webdriver
+import time
+from api.utils.common import Common
+
+class Crawler:
+    def __init__(self):
+        self.base_url = 'https://www.instagram.com'
```
Comment (sundas-arooj) [inline api/services/crawler.py:8]: I choose Instagram to perform crawling

File: api/services/crawler.py [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+        self.common = Common()
+
+    def search_images(self, keyword):
+        search_url = f'{self.base_url}/explore/tags/{keyword}/'
+
+        # Start the seasion using Chrome driver (any driver can be used here)
+        driver = webdriver.Chrome()
+        driver.get(search_url)
```
Comment (sundas-arooj) [inline api/services/crawler.py:16]: I started the driver session, you can use any other driver instead of Chrome

File: api/services/crawler.py [hunk 11]
```diff
@@ -0,0 +1,28 @@
...
+        
+        # Wait for 5 seconds to load the page's content
+        time.sleep(5)
+
+        soup = BeautifulSoup(driver.page_source, 'html.parser')
+        image_elements = soup.find_all('img')
```
Comment (sundas-arooj) [inline api/services/crawler.py:22]: Here I parse the HTML and get all the `img` tags on page using `BeautifulSoup`

File: api/services/classifier.py [hunk 12]
```diff
@@ -0,0 +1,39 @@
+import requests
+import io
+import tensorflow as tf
+import numpy as np
+from PIL import Image
+from api.utils.common import Common
+
+class Classifier:
+    def __init__(self):
+        self.model = tf.keras.applications.mobilenet_v2.MobileNetV2(weights='imagenet', include_top=True, input_shape=(224, 224, 3))
```
Comment (sundas-arooj) [inline api/services/classifier.py:10]: Here I initialize the model MobileNetV2 to perform analysis

File: api/services/classifier.py [hunk 13]
```diff
@@ -0,0 +1,39 @@
...
+            return {'error': 'Failed to load image.'}
+
+        img = tf.keras.applications.mobilenet_v2.preprocess_input(img)
+        predictions = self.model.predict(np.expand_dims(img, axis=0))
+        decoded_predictions = tf.keras.applications.imagenet_utils.decode_predictions(predictions)
+
+        analyzed_results = [{'image': self.common.get_filename(image_path), 'label': label, 'probability': float(prob)} for _, label, prob in decoded_predictions[0]]
+        return analyzed_results
+    
+    def load_image(self, image_url_or_path):
```
Comment (sundas-arooj) [inline api/services/classifier.py:26]: This method will load the image into image array for preprocessing

File: api/services/classifier.py [hunk 14]
```diff
@@ -0,0 +1,39 @@
...
+    
+    def analyze_image(self, image_path):
+        img = self.load_image(image_path)
+
+        if img is None:
+            return {'error': 'Failed to load image.'}
+
+        img = tf.keras.applications.mobilenet_v2.preprocess_input(img)
+        predictions = self.model.predict(np.expand_dims(img, axis=0))
+        decoded_predictions = tf.keras.applications.imagenet_utils.decode_predictions(predictions)
```
Comment (sundas-arooj) [inline api/services/classifier.py:21]: Here model preprocess and perform prediction on the image, It uses the ImageNet dataset for pretraining and image classification.

File: api/utils/common.py
Comment (sundas-arooj) [inline api/utils/common.py:1]: This `utils/common.py` has some common reusable utilities for the project

File: config.py
Comment (sundas-arooj) [inline config.py:1]: This file stores the configuration of the project

File: api/controllers/analyze.py [hunk 15]
```diff
@@ -0,0 +1,31 @@
...
+from api.utils.common import Common
+from config import SUPPORTED_IMAGE_FORMAT
+
+class AnalyzeController(BaseController):
+    def __init__(self):
+        super().__init__()
+        self.classifier = Classifier()
+        self.common = Common()
+
+    def analyze_image(self):
```
Comment (sundas-arooj) [inline api/controllers/analyze.py:13]: This function will be called on calling the endpoint `POST /analyze`

File: api/controllers/analyze.py [hunk 16]
```diff
@@ -0,0 +1,31 @@
...
+        if 'file' in request.files:  # Check if file was uploaded
+            file = request.files['file']
+            if file.filename == '':
+                return self.error_response('No file selected.', 400)
+            elif not self.common.is_file_valid(file):
+                return self.error_response(f'Only {", ".join(SUPPORTED_IMAGE_FORMAT)} image files are supported.', 400)
+        elif 'file' in request.form:  # Check if image URL was provided
+            file = request.form['file']
+        else:
+            return self.error_response('No file found in the request.', 400)
```
Comment (sundas-arooj) [inline api/controllers/analyze.py:23]: Here are some validations on the file, this endpoint can also accept image url

File: api/controllers/analyze.py [hunk 17]
```diff
@@ -0,0 +1,31 @@
...
+
+        # Perform ML analysis on uploaded image
+        analysis = self.classifier.analyze_image(file)
```
Comment (sundas-arooj) [inline api/controllers/analyze.py:26]: Here I perform the analysis on the uploaded file. I called the `analyze_image(file)` function of `api/services/classifier`

File: api/controllers/search_and_analyze.py [hunk 18]
```diff
@@ -0,0 +1,41 @@
...
+        # Get images from Crawler
+        images = self.crawler.search_images(keyword)
+        if not images:
+            return self.error_response('No images found for the keyword.', 404)
+
+        # Perform ML analysis on images
+        analysis_results = []
+        for i, image in enumerate(images):
+            current_time = datetime.now().strftime('%Y-%m-%d_%H-%M-%S')
+            file_path = f'{ASSETS_URL}/{keyword}_{i+1}_{current_time}.jpg'
```
Comment (sundas-arooj) [inline api/controllers/search_and_analyze.py:30]: downloaded files will be stored in `static/assets` folder
